src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1175,7 +1175,27 @@ size_t QgsOgrProvider::layerCount() const
  */
 QgsWkbTypes::Type QgsOgrProvider::wkbType() const
 {
-  return static_cast<QgsWkbTypes::Type>( mOGRGeomType );
+  OGRwkbGeometryType wkb = mOGRGeomType;
+
+  if ( ogrDriverName == QLatin1String( "ESRI Shapefile" ) )
+  {
+    switch ( mOGRGeomType )
```

Comment:
Right that would be a more future proof method. As far as I'm aware, I've only ran into  this issue with shapefiles.